### ISSUE TITLE
pre-commit needs a retry mecanism for network operations

### ISSUE THREAD

**Issue by ssbarnea** [pre-commit/pre-commit#1206]
It seems that pre-commit can fail in CI with what appears to be caused by network operations and it does not do any kind of retry attempt.
```
2019-11-05 09:56:13.953271 | rdo-centos-7 | [INFO] Initializing environment for https://github.com/pre-commit/pre-commit-hooks.
2019-11-05 09:56:16.350589 | rdo-centos-7 | [INFO] Initializing environment for https://github.com/pre-commit/pre-commit-hooks:flake8>=3.7.7.
2019-11-05 09:56:18.641544 | rdo-centos-7 | [INFO] Initializing environment for https://gitlab.com/pycqa/flake8.
2019-11-05 09:56:20.778130 | rdo-centos-7 | An unexpected error has occurred: CalledProcessError: Command: ('/usr/bin/git', 'fetch', 'origin', '--tags')
2019-11-05 09:56:20.778255 | rdo-centos-7 | Return code: 128
2019-11-05 09:56:20.778375 | rdo-centos-7 | Expected return code: 0
2019-11-05 09:56:20.778429 | rdo-centos-7 | Output: (none)
2019-11-05 09:56:20.778462 | rdo-centos-7 | Errors:
2019-11-05 09:56:20.778543 | rdo-centos-7 |     error: RPC failed; result=22, HTTP code = 4042019-11-05 09:56:20.778639 | rdo-centos-7 |     fatal: The remote end hung up unexpectedly
```

Full log at: http://logs.rdoproject.org/91/23591/2/check/tox-linters/f9bf1d6/job-output.txt

**Comment by asottile** [comment]
please consider using softer language ("needs") when making reports

404 isn't a retryable error, this is due to old git on centos

the underlying tools (pip, git, etc.) already manage their own retry protocols -- adding more retries isn't going to help

**Comment by ssbarnea** [comment]
I raised this bug as it was found by two of my colleagues, I knew that faced it myself few months back.

I do agree with you that 404 is not retriable and in 99/100 cases a client should respect it. Still I remember few cases where hosted services returned invalid HTTP responses and where the only option was to add some special cases (Atlassian and their CDN...).

Still the real bug behind my original request would be better described as:

- pre-commit fails to clone flake8 on centos-7 using [officially documented hook configuration](https://flake8.pycqa.org/en/latest/user/using-hooks.html).
- The is a workaround for this case, which is to add the `.git` suffix to the url. Should we just update flake8 doc to suggest something that would work for most people?
- I know that git version on centos-7 is bit ancient (`1.8.3.1`) but fully outside my power to upgrade it or to influence that due to how [LTS support](https://en.wikipedia.org/wiki/CentOS#End-of-support_schedule) works.

As I already found another important bug related to git version: it can fail to checkout when using non-tags on rev field.

PS. Indeed, I will try to soften them a little bit.
